src/core/net/ip6.cpp

```diff
@@ -1201,11 +1205,9 @@ otError Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkM
     {
         if (nextHeader == kProtoIp6)
         {
-            // Remove encapsulating header.
+            // Remove encapsulating header and start over.
             aMessage.RemoveHeader(aMessage.GetOffset());
-
-            IgnoreError(HandleDatagram(aMessage, aNetif, aLinkMessageInfo, aFromNcpHost));
-            ExitNow(tunnel = true);
+            goto start;
```

Comment:
Yes. That's possible. The downside is that it adds another indentation tab to the whole block. Also we need a `break/ExitNow()` at the end of `while` block (also we need to ensure that there is no other `while/for` loop between the top level `while()` and where we want to use `continue` to jump to start- which is the the case here). Practically the `while` loop would be a way to enable `goto start` without using `goto`.

I agree that `goto` should avoided and the main rational for avoiding  `goto` is that it can make the code harder to read and follow but I feel the actual `goto start` in this case may be simpler/easier to read and understand than the `while()` and `continue` model.

We have other example of the same pattern where we use `goto start` in  `MeshForwarder::PrepareDataFrame()`.